Fortios ipv4 policy

bjolivot opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
fortios_ipv4_policy

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (fortios_ipv4_policy a1a1fc6d4b) last updated 2017/02/23 15:56:28 (GMT +200)
  config file = /home/ben/.ansible.cfg
  configured module search path = Default w/o overrides

```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Manage fortigate ipv4 policy objects

bjolivot committed:
New module fortios_address_group

bjolivot committed:
New module fortios_ipv4_policy

bjolivot committed:
New module fortios_ipv4_policy

bjolivot committed:
Fix pep8

bjolivot committed:
Fix alias doc problem

bjolivot commented:
ready_for_review

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+author: "Benjamin Jolivot (@bjolivot)"
+short_description: Manage fortios firewall ipv4 policy objects
+description:
+  - This module provide management of firewall ipv4 policies on FortiOS devices.
```
provides

dagwieers commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+    result = dict(changed=False)
+
+    # fail if libs not present
+    msg = ""
```
Unused, can be removed AFAICT

dagwieers commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+    try:
+        f.open()
+    except:
+        module.fail_json(msg='Error connecting device')
```
Not sure what exceptions could happen, but it may be easier for the end-user to debug if the exception is being reported to.

```python

except:
    e = get_exception()
    module.fail_json(msg='Error connecting device. %s' % e)
```

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+        f.load_config(path=path)
+        result['firewall_address_config'] = f.running_config.to_text()
+    except:
+        module.fail_json(msg='Error reading running config')
```
Call `f.close()` in case of failure (the API doesn't provide a context manager).

dagwieers replied:
Not sure what exceptions could happen, but it may be easier for the end-user to debug if the exception is being reported to.
```python
except:
    e = get_exception()
    module.fail_json(msg='Error connecting device. %s' % e)
```

dagwieers commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+            result['changed'] = True
+
+    #Commit if not check mode
+    if module.check_mode is False and change_string != "":
```
I prefer `if not module.check_mode`

pilou- replied:
I prefer `if not module.check_mode and change_string:`.

dagwieers commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+        except FailedCommit:
+            #rollback
+            e = get_exception()
+            module.fail_json(msg="Unable to commit change, check your args, the error was {0}".format(e.message))
```
Using format() will ensure it fails to work on older python versions.
So I would suggest using the older syntax if there is no real need to do this.

bjolivot replied:
I find it coder's choice, will talk about that on IRC

dagwieers replied:
Well, as long as we support Python 2.6 and older on nodes (so still a couple of years), this change will be required. http://docs.ansible.com/ansible/intro_installation.html#managed-node-requirements

Unless this module cannot work on Python 2.6 and older for other reasons that are not easy to fix. https://pyformat.info/

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+    #connect
+    try:
+        f.open()
+    except:
```
Avoid to catch all exceptions ([see](https://docs.python.org/2/howto/doanddont.html#except)), use at least `Exception` (the same applies below).

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+    if module.params['state'] == 'absent':
+        f.candidate_config[path].del_block(module.params['id'])
+        change_string = f.compare_config()
+        if change_string != "":
```
I prefer `if change_string:`

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+        new_policy = FortiConfig(module.params['id'], 'edit')
+
+        #src / dest / service / interfaces
+        new_policy.set_param('srcintf', '"{0}"'.format(module.params['src_intf']))
```
It could be better to rename `src_intf` and `dst_intf` module parameters reusing names of the API (`srcintf` and `dstintf` respectively). The same for `*addr` parameters below.

bjolivot replied:
My first daft was without "_" but I made lot of errors in playbook, I think it's more readable this way

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+        #check if change needed
+        change_string = f.compare_config()
+
+        if change_string != "":
```
I prefer `if change_string:`.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+        try:
+            f.commit()
+        except FailedCommit:
+            #rollback
```
What does this comment mean ?
I am not sure if `f.rollback()` should be called or not here:
1. In [FortiOS._commit method](https://github.com/spotify/pyfg/blob/master/pyFG/fortios.py#L280), the pyFG library wraps use the `batch` command in order to wrap the commands
2. The [FortiOS cli](http://docs.fortinet.com/uploaded/files/2194/fortiswitchos-cli-30.pdf) (page 91) doesn't document the behavior when a batch command failed.
3. but the wrapped command are executed using `FortiOS.execute_command` which in error cases [raise `CommandExecutionException` exceptions](https://github.com/spotify/pyfg/blob/master/pyFG/fortios.py#L160)
4. These exception aren't caught by the API

Given the doubts, a rollback could be better.

bjolivot replied:
pyFG auto start rollback in case of failure
https://github.com/spotify/pyfg/blob/master/pyFG/fortios.py line 330.
(allready tested a lot while coding ^^)

pilou- replied:
`rollback` method won't be called when `CommandExecutionException` is raised.

bjolivot replied:
That's what i see in pyFG but not what I see when testing.
It seems that there can't be FailedCommitException whitout CommandExecutionException but I have seen a lot of FailedCommit Exception and no  CommandExecutionException

Don't know what to do.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+
+def main():
+    argument_spec = dict(
+        comment                   = dict(),
```
The default type is `str` but it could be added for the `comment` parameter in order te be consistent with other `str` parameter declaration (`poolname`, `av_profile`, `webfilter_profile` ...).

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+try:
+    from pyFG import FortiOS, FortiConfig
+    from pyFG.fortios import logger
+    from pyFG.exceptions import CommandExecutionException, FailedCommit, ForcedCommit
```
`ForcedCommit` isn't used (neither by this module, nor by pyFG library).

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+#check for pyFG lib
+try:
+    from pyFG import FortiOS, FortiConfig
+    from pyFG.fortios import logger
```
`logger` seems imported but not used.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+
+    #merge global required_if & argument_spec from module_utils/fortios.py
+    argument_spec.update(fortios_argument_spec)
+    required_if = fortios_required_if
```
This statement seems useless.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,324 @@
+            new_policy.set_param('nat', 'enable')
+            if module.params['fixedport']:
+                new_policy.set_param('fixedport', 'enable')
+            if module.params['poolname'] is not None:
```
Should not `fixedport` or `poolname` be required when `nat` parameter is set ?

bjolivot replied:
No.
But I need to test that nat is true when fixedport (or poolname) is set.
Will see how to do that

bjolivot committed:
Fix string format for 2.5 compat + close cnx

bjolivot committed:
Forgoten if string != ""

bjolivot commented:
ready_for_review

resmo commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,335 @@
+  state:
+    description:
+      - Specifies if address need to be added or deleted.
+    required: true
```
according to arg spec the default is `present` and as it has a default, it is not required.

resmo commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,335 @@
+      - Negate destination address param.
+    default: false
+    choices: ["true", "false"]
+  action:
```
hmm I really don't like this param name `action`.

`action` is kind of ambiguous, it would rather prefer `policy` or even `action_policy`. Any thoughts?

bjolivot committed:
Fix doc, change action to policy_action & add action as alias

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+        f.candidate_config[path][module.params['id']] = new_policy
+
+        #check if change needed
+        change_string = f.compare_config()
```
Call to `compare_config` and the 2 following statements are duplicated when `state` is equal to `present` or `absent`: move these statements after the conditional blocks.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+            result['changed'] = True
+
+    #Commit if not check mode
+    if change_string and not module.check_mode:
```
There is some code duplicated between this pull-request `fortios_ipv4_policy` and other FortisOS pull-requests ([`fortios_address_group`](#21831) and [`fortios_address`](#21542)):

- block calling `compare_config` method
- block calling `commit` method

These blocks must be moved together in [`module_utils/fortios.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/fortios.py).

Either an helper method or a parent class (see for example [cloudstack](https://github.com/ansible/ansible/tree/devel/lib/ansible/modules/cloud/cloudstack) modules) must be used instead of duplicating the common code.

bjolivot replied:
It could have help finding theses modules at the beginning, will wait for the V2 wich will get rid of pyfg library

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+            result['changed'] = True
+
+    #Present state
+    if module.params['state'] == 'present':
```
Use `elif` instead of `if`.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+#!/usr/bin/python
+#
+# Ansible module to manage IPV4 policy objects in fortigate devices
+# (c) 2016, Benjamin Jolivot <bjolivot@gmail.com>
```
Update the date.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+#!/usr/bin/python
+#
+# Ansible module to manage IPV4 policy objects in fortigate devices
```
Use same case in all comments: `IPv4`.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+    default: 'always'
+  nat:
+    description:
+      - Enable or disable Nat.
```
Use one case in all comments: `NAT`.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+    f = FortiOS( module.params['host'],
+        username=module.params['username'],
+        password=module.params['password'],
+        timeout=module.params['username'],
```
`timeout` and `username` aren't the same thing.

bjolivot committed:
fix doc + bug in timeout + duplicate code for config compare

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,336 @@
+
+    path = 'firewall policy'
+
+    #connect
```
There is some code duplicated between this pull-request `fortios_ipv4_policy` and other FortisOS pull-requests ([`fortios_address_group`](#21831) and [`fortios_address`](#21542)):

- block calling `connect` method
- block calling `load_config` method

These blocks must be moved together in [`module_utils/fortios.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/fortios.py).

Either an helper method or a parent class (see for example [cloudstack](https://github.com/ansible/ansible/tree/devel/lib/ansible/modules/cloud/cloudstack) modules) must be used instead of duplicating the common code.

bjolivot commented:
Wait before review I will release common code in module_utils

bjolivot committed:
Create class AnsibleFortios in module_utils/forios.py + use in ipv4_policy module

bjolivot commented:
As pointed out by @pilou- some code where duplicated between fortios_modules
I created a new class, based on cloustack helper
I add error explainer

ready_for_review

resmo commented on lib/ansible/module_utils/fortios.py:
```diff
@@ -28,8 +28,21 @@
 #
 import os
 import time
+import json
```
seems unused

bjolivot committed:
remove json import

bjolivot commented:
ready_for_review

pilou- commented on lib/ansible/module_utils/fortios.py:
```diff
@@ -58,3 +76,94 @@ def backup(module,running_config):
+
+    def get_error_infos(self, cli_errors):
+        error_list = []
+        for errors in cli_errors:
```
In Python 3, exceptions aren't iterable ([see](http://python3porting.com/noconv.html#handling-exceptions)), use `args` attribute.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,292 @@
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
+
+#check for pyFG lib
```
Remove the `try` block, it is unused.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,292 @@
+    host: 192.168.0.254
+    username: admin
+    password: password
+    id: 42
```
1. I guess examples should use a string for the value of `id` parameter.
2. `name` should be used instead of `id` ([see](http://docs.ansible.com/ansible/dev_guide/developing_modules_best_practices.html)).

bjolivot replied:
id really is an int but need to be a string for pyFG to work, i made changes.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,292 @@
+    #NAT related
+    if not module.params['nat']:
+        if module.params['poolname']:
+            module.fail_json(msg='Poolname param require NAT to be true.')
```
Typo: `requires`.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,292 @@
+
+    #test params
+    #NAT related
+    if not module.params['nat']:
```
It would be better if `AnsibleFortios()` was called before: availability of the dependency (pyfg) should be tested first.

pilou- commented on lib/ansible/modules/network/fortios/fortios_ipv4_policy.py:
```diff
@@ -0,0 +1,292 @@
+
+
+    #init forti_device object
+    forti_device = AnsibleFortios(module)
```
It would be better to use another name, `forti_device` is already used in `module_utils/fortios.py` with another type.

bjolivot committed:
python3 error handling compatibility
bad examples for srcadd or dstaddr s/any/all/
remove pyFG dependency in module (moved to module_utils)
id type is int but casted as string
call fortiosansible object sooner
typo in doc

resmo commented:
shipit
